### PR TITLE
Improve training

### PR DESCRIPTION
**Pull request by jmarshrossney**
Implements main improvements to training:

- Remove ReLU on final **t** network layer (also structures networks into blocks of linear+activation)
- Scheduler parameters specified in runcard (patience and learning rate reduction factor being the most crucial)
- Optimizer and optimizer parameters in runcard (Adadelta / Adam currently appear to be the best options for single layer / multiple layer networks)

To do
- [ ] Allow complete specification of network in runcard
- [ ] Better way of linking optimizer runcard parameter to torch.optim functions and kwargs
- [ ] Find a way around setting the default patience in the config parser

This branch is based on uncertainties (f47da81).
I realise I'm jumping the gun with this PR, but hopefully it can be merged quickly after uncertainties because they're all minor changes.
Would it be easier to view changes if I set uncertainties as the base branch i.e. branch to be merged with?

### CODE REVIEW

**Comment by wilsonmr** [comment]
I'm just changed the base branch so I can see relevant changes. We will change this back once #23 is merged

**Comment by wilsonmr** [comment]
on surface looks pretty good! Like you say we shoulod merge other branch first.

Personally I really dislike sectioning the files with comments, it seems really unneccessary to me and is sufficient to just group relevant dependencies together. If you are going to do it can you at least use a consistent style? Maybe it will change my mind...

to be specific:

```python
#######################################
###     Bootstrap distributions     ###
#######################################
```

```python
# --- Action --- #
```

choose one or get rid of both

**Comment by wilsonmr** [comment]
looks great, can you remind me why we needed the block stuff in the network?

File: anvil/observables.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import torch
 from tqdm import tqdm
 
+
```
Comment (wilsonmr) [inline anvil/observables.py]: I think we will just resolve the conflict by abandoning the changes to observable

**Comment by wilsonmr** [comment]
tempted to make the runcard flat, instead of having kwarg dicts, just spec the kwargs and put them together.. I'll think about this

File: anvil/observables.py (hunk 1 above)
Comment (wilsonmr) [reply anvil/observables.py]: in fact all of the formatting changes which I guess were made by `black` we will probbaly abandon here, since otherwise the files are untouched

**Comment by jmarshrossney** [comment]
> looks great, can you remind me why we needed the block stuff in the network?

No particularly good reason other than I found it easy to mess with `_block` during testing than directly with the `nn.ModuleList`s.



> tempted to make the runcard flat, instead of having kwarg dicts, just spec the kwargs and put them together.. I'll think about this

Absolutely fine by me. I was having trouble with optional parameters (optimiser and scheduler kwargs) in the parser, and specifying a dict was the neatest way I could find to get round this.

**Comment by wilsonmr** [comment]
feel free to make these changes

**Comment by jmarshrossney** [comment]
Changed base branch back to master after merging more recent changes, and ran the example phi^4 train and sample scripts.

Leaning towards leaving the "complete specification of network[s] in runcard" until a later PR, since they don't require tuning for the phi^4 theory.

**Comment by wilsonmr** [comment]
> Leaning towards leaving the "complete specification of network[s] in runcard" until a later PR, since they don't require tuning for the phi^4 theory.

good idea! LGTM

**Review by wilsonmr — APPROVED** [review]
Looks fantastic, thanks for adding examples!
